Changes to versions for 3.12 to work

Chrillebon opened the pull request:
Notes on changes

- numpy stays at version at version 1.24, but with python 3.12, should automatically pick 1.26 (which is supported as stated [here](https://numpy.org/news/#numpy-1260-released)).
- Allowing sympy-plot-backends to extend up to before 3.0. Version 3.0 and onwards have breaking changes with dtumathtools.
- Reflected changes in packages at [conda-forge](https://github.com/Chrillebon/dtumathtools-feedstock). When merging/changing this branch, it should be reflected there as well. Can only be completed once package is uploaded to pypi, as checks that are being run include downloading the new version from pypi. *Note: changed to a fork and not a branch.*
- Updated version label to 2024.1.0
- Added python 3.12 to workflow.
- Passes all tests (in python 3.12.1). However, pytest has 1 warning related to datetime and python 3.12 in a dateutil package. Comes from sympy-plot-backends, and is thus not a direct result of this package. This is not considered major, and can be ignored.
- Mayavi 4.8.2 has not yet been released. Thus to get dtumathtools onto pypi, the install in pyproject.toml needs to be removed (cannot have install based on path in pypi). Another PR will remove support for mayavi for the spring 24 release.

Chrillebon commented:
All tests passed except the ones with python 3.8. This is due to python 3.8 not being supported by numpy 1.26 (which is required to support python 3.12). 

"Simplest solution" would be to drop support for python 3.8.

zerothi commented on pyproject.toml:
```diff
@@ -48,16 +48,16 @@ build-backend = "setuptools.build_meta"
     "sympy~=1.12", 
     "matplotlib==3.6.*", 
     "jupyter>=1.0.0", 
-    "numpy>=1.24",
```
I don't think we need to bump numpy requirement, for Py 3.12, it will automatically grab 1.26, so I don't think it should be a problem, no?
Keep it as is if the tests succeeds.

Chrillebon replied:
Good point! I was afraid that people that already had np 1.24 and python 3.12 would have problems if we did not bump, but as this is not possible, I do not think we should have any problems with `numpy>=1.24`. 

I will revert the change.

zerothi commented on pyproject.toml:
```diff
@@ -48,16 +48,16 @@ build-backend = "setuptools.build_meta"
+    "numpy>=1.26",
 ]
-requires-python = ">=3.8,<3.12"
+requires-python = ">=3.8"
```
Should we do `<3.13`?

Chrillebon replied:
Yes! Absolutely!

zerothi commented:
> Notes on changes
> 
>     * Bump numpy version to 1.26. As stated [here](https://numpy.org/news/#numpy-1260-released), this version supports python 3.9-3.12.

I think we should stick with 3.8 as well, jumping across *still supported* Python versions is bad in this case IMHO.
> 
>     * Allowing sympy-plot-backends to extend up to before 3.0. Version 3.0 and onwards have breaking changes with dtumathtools.

Good idea, keep it like this until autumn semester. Then we can bump it.

> 
>     * Reflected changes in packages at [conda-forge](https://github.com/conda-forge/dtumathtools-feedstock/tree/package_versions_python312). When merging/changing this branch, it should be reflected there as well.
> 
>     * Updated version label to 2024.1.0
> 
>     * Added python 3.12 to workflow.
> 
>     * Passes all tests (in python 3.12.1). However, pytest has 1 warning related to datetime and python 3.12 in a dateutil package. Comes from sympy-plot-backends, and is thus not a direct result of this package. This is not considered major, and can be ignored.
> 
>     * Mayavi 4.8.2 has not yet been released. Thus to get dtumathtools onto pypi, the install in pyproject.toml needs to be removed (cannot have install based on path in pypi). We might consider simply removing support for mayavi for the spring 24 release?

It shouldn't be required to be removed from the `pyproject.toml` file. Students have been asked to do `pip install dtumathtools` which will only use the `project.dependencies` block. The `mayavi` dependency will only get installed when doing: `pip install dtumathtools[mayavi]`.

Hence it hasn't been installed anyways.

Chrillebon commented:
No, I think you misunderstand what I meant. dtumathtools cannot be uploaded to PyPi with such a dependency (using a git path). Therefore, the package currently on pypi does not have `mayavi` supported (pip install dtumathtools[mayavi] does not install mayavi). I think we probably should reflect that change in the github, explicitly stating that mayavi is not supported. It is currently not the case, such that `pip install .[mayavi]` (after having downloaded the github package) still works with mayavi. 
As it is not part of the course, it should not be a major problem though.

zerothi commented:
> No, I think you misunderstand what I meant. dtumathtools cannot be uploaded to PyPi with such a dependency (using a git path). Therefore, the package currently on pypi does not have `mayavi` supported (pip install dtumathtools[mayavi] does not install mayavi). I think we probably should reflect that change in the github, explicitly stating that mayavi is not supported. It is currently not the case, such that `pip install .[mayavi]` (after having downloaded the github package) still works with mayavi. As it is not part of the course, it should not be a major problem though.

Ok, I would be ok with leaving out `mayavi` from supported backends. Could you create a new PR for this?

Chrillebon commented:
I have removed support for `mayavi` in a PR [here](https://github.com/dtudk/dtumathtools/pull/25).

Notes on this PR: I updated the description so it should reflect the current changes. Furthermore, all tests successfully passed, so it should be ready for your review, @zerothi.
